### PR TITLE
Use next free port in FlapdoodleMongoTestConnectionProviderSource

### LINKED ISSUE #2860: More robust port detection in FlapdoodleMongoTestConnectionProviderSource
`FlapdoodleMongoTestConnectionProviderSource` currently determines the port that Mongo is running on by looking at the log output of the mongo daemon and checking for a particular regex.
When upgrading flapdoodle from `3.1.4` to `3.2.2` in #2840 the logging pattern changed and so the regex had to be updated.
It would be good to make that port detection code more robust by e.g. avoiding log parsing

### PR DESCRIPTION
**Pull request by nastra**
fixes #2860

This uses `ServerSocket` to retrieve the next available port in random
fashion and pass it to the `MongodConfig` rather than having to parse
the log output to figure out on which random port Mongo itself is
running.

Flapdoodle itself is doing a similar approach in https://github.com/flapdoodle-oss/de.flapdoodle.embed.process/blob/master/src/main/java/de/flapdoodle/embed/process/runtime/Network.java#L134-L155

### CODE REVIEW

File: versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java [hunk 1]
```diff
@@ -120,4 +80,10 @@ public void stop() throws Exception {
       }
     }
   }
+
+  private Integer findRandomOpenPortOnAllLocalInterfaces() throws IOException {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      return socket.getLocalPort();
```
Comment (dimas-b) [inline versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: Technically this does not guarantee that Mongo will be able to bind to this port later. Is monitoring Mongo logs that much of an inconvenience?

Comment (nastra) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: monitoring the mongo log depends on the logging format :shrug:

Comment (dimas-b) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: true, but it's a solid test failure when it changes vs. a random failure with port allocation in odd environments... :shrug: 

@snazy : What is your take of this? I believe we talked briefly about this before :)

Comment (nastra) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: alternatively I was thinking whether there's a way to ask flapdoodle for the port that mongo is running on

Comment (dimas-b) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: May a PR to Flapdoodle? ;-)

Comment (snazy) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/FlapdoodleMongoTestConnectionProviderSource.java:86]: There's no way to let mongod tell you in a machine readable way which port it has chosen.

But Unix domain sockets help here #2863

**Comment by nastra** [comment]
closing in favor of #2863

**Pull request closed without merging**